pkg/distributor/distributor_test.go

```diff
@@ -3722,7 +3845,29 @@ func prepareIngesterZone(zone string, state ingesterZoneState, cfg prepConfig) [
 			labelNamesStreamResponseDelay: labelNamesStreamResponseDelay,
 			timeOut:                       cfg.timeOut,
 			circuitBreakerOpen:            cfg.circuitBreakerOpen,
-		})
+		}
+
+		// Init the partition reader if the ingest storage is enabled.
+		// This is required to let each mocked ingester to consume their own partition.
+		if cfg.ingestStorageEnabled {
+			var err error
+
+			kafkaCfg := ingest.KafkaConfig{}
+			flagext.DefaultValues(&kafkaCfg)
+			kafkaCfg.Address = cfg.ingestStorageKafka.ListenAddrs()[0]
+			kafkaCfg.Topic = kafkaTopic
+			kafkaCfg.LastProducedOffsetPollInterval = 100 * time.Millisecond
+			kafkaCfg.LastProducedOffsetRetryTimeout = 100 * time.Millisecond
+
+			ingester.partitionReader, err = ingest.NewPartitionReaderForPusher(kafkaCfg, ingester.partitionID(), ingester.instanceID(), newMockIngesterPusherAdapter(ingester), log.NewNopLogger(), nil)
+			require.NoError(t, err)
+			require.NoError(t, services.StartAndAwaitRunning(context.Background(), ingester.partitionReader))
```

Comment:
this has the potential to slow down tests significantly if the partition reader suddenly starts doing something time-consuming at startup. WDYT about starting all of them async and then waiting for all of them at the end of `prepareIngesterZone`?

Reply:
Good idea, done in https://github.com/grafana/mimir/pull/7393/commits/19187475187f328a4bb3836e1c0ee298d59b9ce7.